### ISSUE TITLE
HTML in contenteditable divs / rich text editors

### ISSUE THREAD

**Issue by Jab2870** [glacambre/firenvim#812]
- OS Version: linux x86-64
- Browser Version:  Firefox 84.0
- Browser Addon Version: 0.1.32
- Neovim Plugin Version: 0.1.32

This is more a feature request than a bug. When using this in a rich text editor
on the web, it would be nice to be able to explicitly write html in the vim
editor. For example, when writing an email in Outlook, I may wish to make a word
bold.

I tried

```
<b>bold</b>
```

However firevim (I assume) escaped the html so the editor saw

```
&lt;b&gt;bold&lt;/b&gt;
```

I understand that in most cases, this is probably desirable but I would
definitely be interested in a configuration option for this.

Thanks for creating such an amazing tool

**Comment by glacambre** [comment]
Yes, I've been thinking about this for a while. I'm not sure what the best way to offer this feature is though. Do we want a function? Do we want a setting? Do we want both? Something else? This needs a bit of designing :)

**Comment by Jab2870** [comment]
I don't know if it would be possible but I think the easiest way in terms of
configuration would be to have an option in the localSettings setting.

Maybe something like:

```
let g:firenvim_config = { 
    \ 'globalSettings': {
        \ 'alt': 'all',
    \  },
    \ 'localSettings': {
        \ 'mail.*': {
            \ 'escapeHTML': 'no',
        \ },
    \ }
\ }
```

I have no idea if this would be possible in terms of implementation though

**Comment by Jab2870** [comment]
I would be interested in having a go at implementing this in the next couple of
days. However, I cannot seem to find where the HTML encoding is being done.
Could you point me in the right direction?

**Comment by glacambre** [comment]
Thanks a lot for wanting to take the matter in your own hands! :)

You can find instructions about building Firenvim and loading it in your browser [here](https://github.com/glacambre/firenvim#from-source).

First, you will need to declare a new setting in the [configuration type](https://github.com/glacambre/firenvim/blob/03af5cd28d8492943adbea2a2006c359efa5dd1f/src/utils/configuration.ts#L21-L26). Once that's done, you'll need to make sure it has a default value when [loaded from the user's config](https://github.com/glacambre/firenvim/blob/03af5cd28d8492943adbea2a2006c359efa5dd1f/src/background.ts#L178). After that, you will be able to use [configuration.ts:getConf](https://github.com/glacambre/firenvim/blob/03af5cd28d8492943adbea2a2006c359efa5dd1f/src/utils/configuration.ts#L66-L68) in order to fetch the setting in [TextareaEditor.getContent](https://github.com/glacambre/firenvim/blob/03af5cd28d8492943adbea2a2006c359efa5dd1f/src/editors/TextareaEditor.ts#L21) and [TextareaEditor.setContent](https://github.com/glacambre/firenvim/blob/03af5cd28d8492943adbea2a2006c359efa5dd1f/src/editors/TextareaEditor.ts#L55) to check if you need to return the value of [HTMLElement.innerHTML](https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML) rather than that of HTMLElement.innerText.

**Comment by Jab2870** [comment]
Super, thank you. I will have a look into it. It will probably be tomorrow but
I'll certainly do it.

You mentioned you wanted to discuss the best way to do it, are you happy to do
it with the setting as I suggested? Do you think escapeHTML is the best name for
the configuration option?

**Comment by glacambre** [comment]
I'm not sure `escapeHTML` is the best name as there is no escaping taking place. I think a `content` setting whose values could be either `html` or `text` might be better.

**Comment by glacambre** [comment]
Closed in #818, thanks! This feature will be available once I publish a new release.
